core/database_util.go

```diff
@@ -36,34 +36,71 @@ var (
+	headerNumSuffix     = []byte("-num")          // headerPrefix + num (uint64 big endian) + headerNumSuffix -> hash
+	blockHashPrefix     = []byte("blk-hash-")     // blockHashPrefix + hash -> num (uint64 big endian)
+	bodyPrefix          = []byte("body-")         // bodyPrefix + num (uint64 big endian) + hash -> block body
+	blockReceiptsPrefix = []byte("blk-receipts-") // blockReceiptsPrefix + num (uint64 big endian) + hash -> block receipts
 
-	headerSuffix = []byte("-header")
-	bodySuffix   = []byte("-body")
-	tdSuffix     = []byte("-td")
-
-	txMetaSuffix        = []byte{0x01}
-	receiptsPrefix      = []byte("receipts-")
-	blockReceiptsPrefix = []byte("receipts-block-")
+	txMetaSuffix   = []byte{0x01}
+	receiptsPrefix = []byte("receipts-")
 
 	mipmapPre    = []byte("mipmap-log-bloom-")
 	MIPMapLevels = []uint64{1000000, 500000, 100000, 50000, 1000}
 
-	blockHashPrefix = []byte("block-hash-") // [deprecated by the header/block split, remove eventually]
-
 	configPrefix = []byte("ethereum-config-") // config prefix for the db
+
+	// used by old (non-sequential keys) db, now only used for conversion
+	oldBlockPrefix         = []byte("block-")
+	oldHeaderSuffix        = []byte("-header")
+	oldBodySuffix          = []byte("-body")
+	oldBlockNumPrefix      = []byte("block-num-")
+	oldBlockReceiptsPrefix = []byte("receipts-block-")
+	oldBlockHashPrefix     = []byte("block-hash-") // [deprecated by the header/block split, remove eventually]
 )
 
+// encodeBlockNumber encodes a block number as big endian uint64
+func encodeBlockNumber(number uint64) []byte {
+	enc := make([]byte, 8)
+	binary.BigEndian.PutUint64(enc, number)
+	return enc
+}
+
 // GetCanonicalHash retrieves a hash assigned to a canonical block number.
 func GetCanonicalHash(db ethdb.Database, number uint64) common.Hash {
-	data, _ := db.Get(append(blockNumPrefix, big.NewInt(int64(number)).Bytes()...))
+	data, _ := db.Get(append(append(headerPrefix, encodeBlockNumber(number)...), headerNumSuffix...))
 	if len(data) == 0 {
-		return common.Hash{}
+		data, _ = db.Get(append(oldBlockNumPrefix, big.NewInt(int64(number)).Bytes()...))
+		if len(data) == 0 {
+			return common.Hash{}
+		}
 	}
 	return common.BytesToHash(data)
 }
 
+// missingNumber is returned by GetBlockNumber if no header with the
+// given block hash has been stored in the database
+const missingNumber = uint64(0xffffffffffffffff)
+
+// GetBlockNumber returns the block number assigned to a block hash
+// if the corresponding header is present in the database
+func GetBlockNumber(db ethdb.Database, hash common.Hash) uint64 {
+	data, _ := db.Get(append(blockHashPrefix, hash.Bytes()...))
+	if len(data) != 8 {
+		data, _ := db.Get(append(append(oldBlockPrefix, hash.Bytes()...), oldHeaderSuffix...))
+		if len(data) == 0 {
+			return missingNumber
+		}
+		header := new(types.Header)
+		if err := rlp.Decode(bytes.NewReader(data), header); err != nil {
+			return missingNumber
```

Comment:
Preferably do a `glog.Fatalf` here to abort as RLP decoding cannot fail, and if it fails, there's no way to recover.